File: 987_vertical_traversal/vertical_traversal.go
/*
 * @Date: 2021-07-31 00:49:44
 * @Author: Mengsen Wang
 * @LastEditors: Mengsen Wang
 * @LastEditTime: 2021-07-31 01:44:30
 */

package main

import (
	"math"
	"reflect"
	"sort"
)

type TreeNode struct {
	Val   int
	Left  *TreeNode
	Right *TreeNode
}

type data struct{ col, row, val int }

func verticalTraversal(root *TreeNode) (ans [][]int) {
	nodes := []data{}
	var dfs func(*TreeNode, int, int)
	dfs = func(node *TreeNode, row, col int) {
		if node == nil {
			return
		}
		nodes = append(nodes, data{col, row, node.Val})
		dfs(node.Left, row+1, col-1)
		dfs(node.Right, row+1, col+1)
	}
	dfs(root, 0, 0)

	sort.Slice(nodes, func(i, j int) bool {
		a, b := nodes[i], nodes[j]
		return a.col < b.col || a.col == b.col && (a.row < b.row || a.row == b.row && a.val < b.val)
	})

	lastCol := math.MinInt32
	for _, node := range nodes {
		if node.col != lastCol {
			lastCol = node.col
			ans = append(ans, nil)
		}
		ans[len(ans)-1] = append(ans[len(ans)-1], node.val)
	}
	return
}

func main() {
	assert := func(a, b [][]int) {
		if !reflect.DeepEqual(a, b) {
			panic("Not Passed")
		}
	}
	{
		root := &TreeNode{3, &TreeNode{9, nil, nil}, &TreeNode{20, &TreeNode{15, nil, nil}, &TreeNode{7, nil, nil}}}
		ans := [][]int{{9}, {3, 15}, {20}, {7}}
		assert(verticalTraversal(root), ans)
	}
	{
		root := &TreeNode{1, &TreeNode{2, &TreeNode{4, nil, nil}, &TreeNode{5, nil, nil}}, &TreeNode{3, &TreeNode{6, nil, nil}, &TreeNode{7, nil, nil}}}
		ans := [][]int{{4}, {2}, {1, 5, 6}, {3}, {7}}
		assert(verticalTraversal(root), ans)
	}
	{
		root := &TreeNode{1, &TreeNode{2, &TreeNode{4, nil, nil}, &TreeNode{6, nil, nil}}, &TreeNode{3, &TreeNode{5, nil, nil}, &TreeNode{7, nil, nil}}}
		ans := [][]int{{4}, {2}, {1, 5, 6}, {3}, {7}}
		assert(verticalTraversal(root), ans)
	}
}

File: rust/987_vertical_traversal/vertical_traversal.rs
/*
 * @Date: 2021-07-31 00:49:38
 * @Author: Mengsen Wang
 * @LastEditors: 854284842@qq.com
 * @LastEditTime: 2024-02-13
 */

use std::cell::RefCell;
use std::rc::Rc;

// Definition for a binary tree node.
#[derive(Debug, PartialEq, Eq)]
pub struct TreeNode {
    pub val: i32,
    pub left: Option<Rc<RefCell<TreeNode>>>,
    pub right: Option<Rc<RefCell<TreeNode>>>,
}

impl TreeNode {
    #[inline]
    pub fn new(val: i32) -> Self {
        TreeNode {
            val,
            left: None,
            right: None,
        }
    }
}

pub struct Solution {}

impl Solution {
    pub fn vertical_traversal(root: Option<Rc<RefCell<TreeNode>>>) -> Vec<Vec<i32>> {
        if root.is_none() {
            return vec![];
        }

        fn dfs(arr: &mut Vec<(i32, i32, i32)>, cur: &TreeNode, row: i32, col: i32) {
            arr.push((col, row, cur.val));
            if let Some(nd) = cur.left.as_ref() {
                dfs(arr, &nd.borrow(), row + 1, col - 1);
            }
            if let Some(nd) = cur.right.as_ref() {
                dfs(arr, &nd.borrow(), row + 1, col + 1);
            }
        }

        let root = root.unwrap();
        let mut arr = vec![];
        dfs(&mut arr, &root.borrow(), 0, 0);
        arr.sort();
        arr.into_iter()
            .fold((i32::MIN, vec![]), |(prev_c, mut ans), (c, _, v)| {
                if prev_c != c {
                    ans.push(vec![v]);
                } else {
                    ans.last_mut().unwrap().push(v);
                }
                (c, ans)
            })
            .1
    }
}

fn main() {
    let tests = vec![
        (
            Some(Rc::new(RefCell::new(TreeNode {
                val: 3,
                left: Some(Rc::new(RefCell::new(TreeNode {
                    val: 9,
                    left: None,
                    right: None,
                }))),
                right: Some(Rc::new(RefCell::new(TreeNode {
                    val: 20,
                    left: Some(Rc::new(RefCell::new(TreeNode {
                        val: 15,
                        left: None,
                        right: None,
                    }))),
                    right: Some(Rc::new(RefCell::new(TreeNode {
                        val: 7,
                        left: None,
                        right: None,
                    }))),
                }))),
            }))),
            vec![vec![9], vec![3, 15], vec![20], vec![7]],
        ),
        (
            Some(Rc::new(RefCell::new(TreeNode {
                val: 1,
                left: Some(Rc::new(RefCell::new(TreeNode {
                    val: 2,
                    left: Some(Rc::new(RefCell::new(TreeNode {
                        val: 4,
                        left: None,
                        right: None,
                    }))),
                    right: Some(Rc::new(RefCell::new(TreeNode {
                        val: 5,
                        left: None,
                        right: None,
                    }))),
                }))),
                right: Some(Rc::new(RefCell::new(TreeNode {
                    val: 3,
                    left: Some(Rc::new(RefCell::new(TreeNode {
                        val: 6,
                        left: None,
                        right: None,
                    }))),
                    right: Some(Rc::new(RefCell::new(TreeNode {
                        val: 7,
                        left: None,
                        right: None,
                    }))),
                }))),
            }))),
            vec![vec![4], vec![2], vec![1, 5, 6], vec![3], vec![7]],
        ),
        (
            Some(Rc::new(RefCell::new(TreeNode {
                val: 1,
                left: Some(Rc::new(RefCell::new(TreeNode {
                    val: 2,
                    left: Some(Rc::new(RefCell::new(TreeNode {
                        val: 4,
                        left: None,
                        right: None,
                    }))),
                    right: Some(Rc::new(RefCell::new(TreeNode {
                        val: 6,
                        left: None,
                        right: None,
                    }))),
                }))),
                right: Some(Rc::new(RefCell::new(TreeNode {
                    val: 3,
                    left: Some(Rc::new(RefCell::new(TreeNode {
                        val: 5,
                        left: None,
                        right: None,
                    }))),
                    right: Some(Rc::new(RefCell::new(TreeNode {
                        val: 7,
                        left: None,
                        right: None,
                    }))),
                }))),
            }))),
            vec![vec![4], vec![2], vec![1, 5, 6], vec![3], vec![7]],
        ),
    ];

    for (root, ans) in tests {
        assert_eq!(Solution::vertical_traversal(root), ans);
    }
}
